packages/backend/src/queryBuilder.ts

```diff
@@ -108,7 +108,8 @@ export const replaceUserAttributes = (
 
     return sqlAttributes.reduce<string>((acc, sqlAttribute) => {
         const attribute = sqlAttribute.replace(userAttributeRegex, '$1');
-        const userValue: string | null | undefined = userAttributes[attribute];
+        const userValue: string | null | undefined =
+            userAttributes[attribute]?.[0]; // Pick first value, where user value takes precedence over group values
```

Comment:
Update row permission logic to use the first value available.

Reply:
@ZeRego this feels like a fragile api, as a dev how do I know that [0] is the highest precedence attribute?

Somewhere there should be very clear business logic that we have user attributes + group attributes and how we pick the correct one. This approach masks the underlying logic

Reply:
> Somewhere there should be very clear business logic that we have user attributes + group attributes and how we pick the correct one. This approach masks the underlying logic

In that case, the model would expose the default value, user value, groups values and we need to move the logic to utils functions since we will need it in multiple places. 

```
export type UserAttributeValueMap = Record<string, {
    default: string | null;
    user: string | null;
    groups: string[];
}>;
```

Util functions

```
 export const getAllAttributeValues = (userAttribute: UserAttributeValueMap[number]): string[] => {
     let finalValues: string[];
     if (userAttribute.user) {
         finalValues = [userAttribute.user, ...userAttribute.groups];
     } else if (userAttribute.groups.length > 0) {
         finalValues = userAttribute.groups;
     } else if (userAttribute.default) {
         finalValues = [userAttribute.default];
     } else {
         finalValues = [];
     }
     return finalValues;
 }

export const getAttributeValue = (userAttribute: UserAttributeValueMap[number]): string | null => {
    if (userAttribute.user) {
        return userAttribute.user;
    } else if (userAttribute.groups.length > 0) {
        return userAttribute.groups[0];
    } else if (userAttribute.default) {
        return userAttribute.default;
    } else {
        return null;
    }
}
```

For row permission: 

```
const userValue: string | null = userAttributes[attribute] ? getAttributeValue(userAttributes[attribute]) : null;
```

for column permissions:

```
export const hasUserAttribute = (
    userAttributes: UserAttributeValueMap,
    attributeName: string,
    value: string,
): boolean =>
    !!userAttributes[attributeName] &&
    getAllAttributeValues(userAttributes[attributeName]).includes(value);
```

Reply:
this work will be done in this ticket [Add reference to all attribute values in row level filtering#8536](https://github.com/lightdash/lightdash/issues/8536)